From 412c130994c72a32780d8285b769dbae8ed1df36 Mon Sep 17 00:00:00 2001
From: Yehuda Yitschak <yehuday@marvell.com>
Date: Thu, 5 May 2016 14:47:01 +0300
Subject: [PATCH 0230/1345] dts: a8040: introduce initial DTS file hierarchy
 for Armada-8040

commit  d52ede3f8143b66591e9060bffff6c4d92e4569e from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

Due to conflicts on clock phandles we cannot re-use armada-cp110.dtsi
file inside armada-cp110-slave.dtsi so we clone it into armada-cp110-1.dtsi

Enabled interfaces on CP1:
- MAC0: 10G(KR)
- SATA1 via comphy#3

Updated interrupt ID's according to CP:
- 64 CP1 interrupts mapping in AP starts in 288
- CP0 interrupts starts in 64, hence the difference in DTS mapping is 224

Change-Id: If00d3b58e4fc15d7ed4dc59e362b31cbd21fb5f7
Signed-off-by: Omri Itach <omrii@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/29715
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../bindings/arm/marvell/armada-7k-8k.txt          |    9 +
 arch/arm64/boot/dts/marvell/Makefile               |    2 +
 .../boot/dts/marvell/armada-8040-db-router.dts     |  149 +++++++++++
 arch/arm64/boot/dts/marvell/armada-8040-db.dtsi    |   71 +++++
 arch/arm64/boot/dts/marvell/armada-8040.dtsi       |   39 +++
 arch/arm64/boot/dts/marvell/armada-cp110-1.dtsi    |  279 +++++++++++++++++++
 .../arm64/boot/dts/marvell/armada-cp110-slave.dtsi |  281 ++++++++++++++++----
 7 files changed, 779 insertions(+), 51 deletions(-)
 create mode 100644 arch/arm64/boot/dts/marvell/armada-8040-db-router.dts
 create mode 100644 arch/arm64/boot/dts/marvell/armada-8040-db.dtsi
 create mode 100644 arch/arm64/boot/dts/marvell/armada-cp110-1.dtsi

diff --git a/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt b/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt
index f8a725c..1f7ed03 100644
--- a/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt
+++ b/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt
@@ -6,6 +6,7 @@ the following root node property:
 
  - compatible, with one of the following values:
 
+SoC:
    - "marvell,armada7020", "marvell,armada-ap806-dual", "marvell,armada-ap806"
       when the SoC being used is the Armada 7020
 
@@ -18,6 +19,14 @@ the following root node property:
    - "marvell,armada8040", "marvell,armada-ap806-quad", "marvell,armada-ap806"
       when the SoC being used is the Armada 8040
 
+Boards:
+   - "marvell,armada7040-db-router", "marvell,armada7040-db"
+      when the board being used is the Armada 7040 Development board
+
+   - "marvell,armada8040-db-router", "marvell,armada8040-db"
+      when the board being used is the Armada 8040 Development board
+
+
 Example:
 
 compatible = "marvell,armada7040-db-router", "marvell,armada7040-db", "marvell,armada7040",
diff --git a/arch/arm64/boot/dts/marvell/Makefile b/arch/arm64/boot/dts/marvell/Makefile
index 1279b5c..666fd05 100644
--- a/arch/arm64/boot/dts/marvell/Makefile
+++ b/arch/arm64/boot/dts/marvell/Makefile
@@ -17,6 +17,8 @@ dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-rz-pd.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-db.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-mcbin.dtb
 
+dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-db-router.dtb
+
 always		:= $(dtb-y)
 subdir-y	:= $(dts-dirs)
 clean-files	:= *.dtb
diff --git a/arch/arm64/boot/dts/marvell/armada-8040-db-router.dts b/arch/arm64/boot/dts/marvell/armada-8040-db-router.dts
new file mode 100644
index 0000000..0b3cee03
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-8040-db-router.dts
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C) 2016 Marvell Technology Group Ltd.
+ *
+ * This file is dual-licensed: you can use it either under the terms
+ * of the GPLv2 or the X11 license, at your option. Note that this dual
+ * licensing only applies to this file, and not this project as a
+ * whole.
+ *
+ *  a) This library is free software; you can redistribute it and/or
+ *     modify it under the terms of the GNU General Public License as
+ *     published by the Free Software Foundation; either version 2 of the
+ *     License, or (at your option) any later version.
+ *
+ *     This library is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ * Or, alternatively,
+ *
+ *  b) Permission is hereby granted, free of charge, to any person
+ *     obtaining a copy of this software and associated documentation
+ *     files (the "Software"), to deal in the Software without
+ *     restriction, including without limitation the rights to use,
+ *     copy, modify, merge, publish, distribute, sublicense, and/or
+ *     sell copies of the Software, and to permit persons to whom the
+ *     Software is furnished to do so, subject to the following
+ *     conditions:
+ *
+ *     The above copyright notice and this permission notice shall be
+ *     included in all copies or substantial portions of the Software.
+ *
+ *     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ *     OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ *     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *     OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+ * Device Tree file for Marvell Armada 8040 RZ development board
+ * This board file supports the router configuration of the board
+ */
+
+#include "armada-8040-db.dtsi"
+
+/ {
+	model = "Marvell Armada-8040 development board router setup";
+	compatible = "marvell,armada8040-db-router";
+	#address-cells = <2>;
+	#size-cells = <2>;
+
+	chosen { };
+
+	cp110-master {
+		config-space@f2000000 {
+			sata@540000 {
+				status = "okay";
+			};
+			usb3h0: usb3@500000 {
+				status = "okay";
+			};
+			usb3h1: usb3@510000 {
+				status = "okay";
+			};
+			mdio@12a200 {
+				#address-cells = <1>;
+				#size-cells = <0>;
+				phy2: ethernet-phy@2 {
+					reg = <0>;
+				};
+				phy3: ethernet-phy@3 {
+					reg = <1>;
+				};
+			};
+			gop {
+				emac0: mac0 {
+					phy-mode = "kr"; /* lane-2 */
+					status = "okay";
+				};
+				emac2: mac2 {
+					phy-mode = "rgmii"; /* lane-0 */
+					phy = <&phy2>;
+					status = "okay";
+				};
+				emac3: mac3 {
+					phy-mode = "rgmii"; /* rgmii-1 */
+					phy = <&phy3>;
+					status = "okay";
+				};
+			};
+			ppv22@000000 {
+				eth0: eth0@010000 {
+					status = "okay";
+				};
+				eth1: eth1@020000 {
+					status = "okay";
+				};
+				eth2: eth2@030000 {
+					status = "okay";
+				};
+			};
+			eip197@800000 {
+				status = "okay";
+			};
+		};
+
+		pcie@0x640000 {
+			status = "okay";
+		};
+	};
+	cp110-slave {
+		config-space@f4000000 {
+			sata@540000 {
+				status = "okay";
+			};
+			gop {
+				emac0_1: mac0 {
+					phy-mode = "kr"; /* lane-2 */
+					status = "okay";
+				};
+				emac2_1: mac2 {
+					phy-mode = "rgmii"; /* lane-0 */
+					phy = <&phy2>;
+					status = "disabled";
+				};
+				emac3_1: mac3 {
+					phy-mode = "rgmii"; /* rgmii-1 */
+					phy = <&phy3>;
+					status = "disabled";
+				};
+			};
+			ppv22@000000 {
+				eth0_1: eth0@010000 {
+					status = "okay";
+				};
+				eth1_1: eth1@020000 {
+					status = "disabled";
+				};
+				eth2_1: eth2@030000 {
+					status = "disabled";
+				};
+			};
+		};
+	};
+};
diff --git a/arch/arm64/boot/dts/marvell/armada-8040-db.dtsi b/arch/arm64/boot/dts/marvell/armada-8040-db.dtsi
new file mode 100644
index 0000000..5f4f2ab
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-8040-db.dtsi
@@ -0,0 +1,71 @@
+/*
+ * Copyright (C) 2016 Marvell Technology Group Ltd.
+ *
+ * This file is dual-licensed: you can use it either under the terms
+ * of the GPLv2 or the X11 license, at your option. Note that this dual
+ * licensing only applies to this file, and not this project as a
+ * whole.
+ *
+ *  a) This library is free software; you can redistribute it and/or
+ *     modify it under the terms of the GNU General Public License as
+ *     published by the Free Software Foundation; either version 2 of the
+ *     License, or (at your option) any later version.
+ *
+ *     This library is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ * Or, alternatively,
+ *
+ *  b) Permission is hereby granted, free of charge, to any person
+ *     obtaining a copy of this software and associated documentation
+ *     files (the "Software"), to deal in the Software without
+ *     restriction, including without limitation the rights to use,
+ *     copy, modify, merge, publish, distribute, sublicense, and/or
+ *     sell copies of the Software, and to permit persons to whom the
+ *     Software is furnished to do so, subject to the following
+ *     conditions:
+ *
+ *     The above copyright notice and this permission notice shall be
+ *     included in all copies or substantial portions of the Software.
+ *
+ *     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ *     OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ *     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *     OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+ * Device Tree file for Marvell Armada 8040 Development board platform
+ */
+
+#include "armada-8040.dtsi"
+
+/ {
+	model = "Marvell Armada 8040 development board";
+	compatible = "marvell,armada8040-db", "marvell,armada8040",
+			"marvell,armada-ap806-quad", "marvell,armada-ap806";
+
+	memory@00000000 {
+		device_type = "memory";
+		reg = <0x0 0x0 0x0 0x80000000>;
+	};
+
+	ap806 {
+		config-space@f0000000 {
+			serial@512000 {
+				status = "okay";
+			};
+			i2c@511000 {
+				status = "okay";
+				clock-frequency = <100000>;
+			};
+		};
+	};
+
+};
diff --git a/arch/arm64/boot/dts/marvell/armada-8040.dtsi b/arch/arm64/boot/dts/marvell/armada-8040.dtsi
index 33813a7..ea510da 100644
--- a/arch/arm64/boot/dts/marvell/armada-8040.dtsi
+++ b/arch/arm64/boot/dts/marvell/armada-8040.dtsi
@@ -53,6 +53,45 @@
 	model = "Marvell Armada 8040";
 	compatible = "marvell,armada8040", "marvell,armada-ap806-quad",
 		     "marvell,armada-ap806";
+
+	ap806 {
+		config-space@f0000000 {
+			smmu: iommu@5000000 {
+				mmu-masters = <&usb3h0 0x483>,
+					      <&usb3h1 0x484>;
+			};
+		};
+	};
+
+	cp110-master {
+		config-space@f2000000 {
+			ppv22@000000 {
+				dma-coherent;
+			};
+			usb3h0: usb3@500000 {
+				#stream-id-cells = <1>;
+			};
+			usb3h1: usb3@510000 {
+				#stream-id-cells = <1>;
+			};
+		};
+		pcie@0x600000 {
+			msi-parent = <&gic_v2m0>;
+		};
+		pcie@0x620000 {
+			msi-parent = <&gic_v2m0>;
+		};
+		pcie@0x640000 {
+			msi-parent = <&gic_v2m0>;
+		};
+	};
+	cp110-slave {
+		config-space@f4000000 {
+			ppv22@000000 {
+				dma-coherent;
+			};
+		};
+	};
 };
 
 /* The RTC requires external oscillator. But on Aramda 80x0, the RTC clock
diff --git a/arch/arm64/boot/dts/marvell/armada-cp110-1.dtsi b/arch/arm64/boot/dts/marvell/armada-cp110-1.dtsi
new file mode 100644
index 0000000..703011d
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-cp110-1.dtsi
@@ -0,0 +1,279 @@
+/****************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+/*
+ * Generic Device Tree describing Marvell Armada CP-110 device
+ */
+
+cp110_clk_1: clk@0x440700 {
+	compatible = "marvell,armada-cp110-clock";
+	reg = <0x440700 0x4>;
+	#clock-cells = <1>;
+};
+
+gateclk_1: clock-gating-control@440220 {
+	compatible = "marvell,armada-cp110-gating-clock";
+	reg = <0x440220 0x4>;
+	clocks = <&cp110_clk_1 3>;
+	#clock-cells = <1>;
+};
+
+sata@540000 {
+	compatible = "marvell,armada-3700-ahci";
+	reg = <0x540000 0x30000>;
+	interrupts = <GIC_SPI 287 IRQ_TYPE_LEVEL_HIGH>;
+	clocks = <&gateclk_1 15>, <&gateclk_1 16>;
+	status = "disabled";
+	port_base = <0x10000>;
+	port_offset = <0x10000>;
+};
+
+usb3h0_1: usb3@500000 {
+	compatible = "generic-xhci";
+	reg = <0x500000 0x4000>;
+	dma-coherent;
+	interrupts = <GIC_SPI 286 IRQ_TYPE_LEVEL_HIGH>;
+	clocks = <&gateclk_1 22>, <&gateclk_1 16>;
+	status = "disabled";
+};
+
+usb3h1_1: usb3@510000 {
+	compatible = "generic-xhci";
+	reg = <0x510000 0x4000>;
+	dma-coherent;
+	interrupts = <GIC_SPI 285 IRQ_TYPE_LEVEL_HIGH>;
+	clocks = <&gateclk_1 23>, <&gateclk_1 16>;
+	status = "disabled";
+};
+
+spi@700600 {
+	compatible = "marvell,armada-380-spi";
+	reg = <0x700600 0x50>;
+	#address-cells = <0x1>;
+	#size-cells = <0x0>;
+	cell-index = <0x0>;
+	clocks = <&cp110_clk 3>;
+	status = "disabled";
+};
+
+spi@700680 {
+	compatible = "marvell,armada-380-spi";
+	reg = <0x700680 0x50>;
+	#address-cells = <0x1>;
+	#size-cells = <0x0>;
+	cell-index = <0x0>;
+	clocks = <&cp110_clk 3>;
+	status = "disabled";
+};
+
+i2c@701000 {
+	compatible = "marvell,mv64xxx-i2c";
+	reg = <0x701000 0x20>;
+	#address-cells = <1>;
+	#size-cells = <0>;
+	interrupts = <GIC_SPI 310 IRQ_TYPE_LEVEL_HIGH>;
+	timeout-ms = <1000>;
+	clocks = <&cp110_clk 3>;
+	status = "disabled";
+};
+
+i2c@701100 {
+	compatible = "marvell,mv64xxx-i2c";
+	reg = <0x701100 0x20>;
+	#address-cells = <1>;
+	#size-cells = <0>;
+	interrupts = <GIC_SPI 311 IRQ_TYPE_LEVEL_HIGH>;
+	timeout-ms = <1000>;
+	clocks = <&cp110_clk 3>;
+	status = "disabled";
+};
+
+nand@720000 {
+	compatible = "marvell,armada370-nand";
+	reg = <0x720000 0x54>;
+	#address-cells = <1>;
+	#size-cells = <1>;
+	num-cs = <1>;
+	marvell,nand-enable-arbiter;
+	nand-on-flash-bbt;
+	interrupts = <GIC_SPI 307 IRQ_TYPE_LEVEL_HIGH>;
+	clocks = <&gateclk_1 2>;
+	nand-ecc-strength = <4>;
+	nand-ecc-step-size = <512>;
+	status = "disabled";
+};
+
+cpxor0@6a0000 {
+	compatible = "marvell,mv-xor-v2";
+	reg = <0x6a0000 0x1000>,
+	      <0x6b0000 0x1000>;
+	dma-coherent;
+	msi-parent = <&gic_v2m0>;
+	clocks = <&gateclk_1 14>, <&gateclk_1 8>;
+};
+
+cpxor1@6c0000 {
+	compatible = "marvell,mv-xor-v2";
+	reg = <0x6c0000 0x1000>,
+	      <0x6d0000 0x1000>;
+	dma-coherent;
+	msi-parent = <&gic_v2m0>;
+	clocks = <&gateclk_1 14>, <&gateclk_1 7>;
+};
+
+mdio@12a200 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	compatible = "marvell,orion-mdio";
+	reg = <0x12a200 0x10>;
+};
+
+mdio@15b000 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	compatible = "marvell,xmdio";
+	reg = <0x12a600 0x10>;
+};
+
+axim-cp-rd@3c5000 {
+	compatible = "marvell,coresight-axim", "arm,primecell";
+	reg = <0x3c5000 0x1000>;
+	clocks = <&syscon 3>;
+	clock-names = "apb_pclk";
+	bus-width = <40>;
+};
+
+axim-cp-wr@3c6000 {
+	compatible = "marvell,coresight-axim", "arm,primecell";
+	reg = <0x3c6000 0x1000>;
+	clocks = <&syscon 3>;
+	clock-names = "apb_pclk";
+	bus-width = <40>;
+};
+
+axim-ppv2-rd@3c0000 {
+	compatible = "marvell,coresight-axim", "arm,primecell";
+	reg = <0x3c0000 0x1000>;
+	clocks = <&syscon 3>;
+	clock-names = "apb_pclk";
+	bus-width = <40>;
+};
+
+axim-ppv2-wr@3c1000 {
+	compatible = "marvell,coresight-axim", "arm,primecell";
+	reg = <0x3c1000 0x1000>;
+	clocks = <&syscon 3>;
+	clock-names = "apb_pclk";
+	bus-width = <40>;
+};
+
+axim-hb1-rd@3c8000 {
+	compatible = "marvell,coresight-axim", "arm,primecell";
+	reg = <0x3c8000 0x1000>;
+	clocks = <&syscon 3>;
+	clock-names = "apb_pclk";
+	bus-width = <40>;
+};
+
+axim-hb1-wr@3c9000 {
+	compatible = "marvell,coresight-axim", "arm,primecell";
+	reg = <0x3c9000 0x1000>;
+	clocks = <&syscon 3>;
+	clock-names = "apb_pclk";
+	bus-width = <40>;
+};
+
+aliases {
+	ethernet0_1 = &emac0_1;
+	ethernet1_1 = &emac2_1;
+	ethernet2_1 = &emac3_1;
+};
+
+gop {
+	emac0_1: mac0 {
+		interrupts = <GIC_SPI 292 IRQ_TYPE_LEVEL_HIGH>;  /* Link IRQ */
+		mac-address = [00 00 00 00 00 04];
+		port-id = <0>; /* gop_port_id */
+	};
+	emac2_1: mac2 {
+		interrupts = <GIC_SPI 291 IRQ_TYPE_LEVEL_HIGH>;  /* Link IRQ */
+		mac-address = [00 00 00 00 00 05];
+		port-id = <2>; /* gop_port_id */
+	};
+	emac3_1: mac3 {
+		interrupts = <GIC_SPI 290 IRQ_TYPE_LEVEL_HIGH>;  /* Link IRQ */
+		mac-address = [00 00 00 00 00 06];
+		port-id = <3>; /* gop_port_id */
+	};
+};
+
+ppv22@000000 {
+	compatible = "marvell,mv-pp22";
+	reg = <0x000000 0x80000>, /* Packet Processor regs */
+	      <0x120000 0x9000>, /* SERDES regs */
+	      <0x129000 0x0600>, /* XMIB regs */
+	      <0x12a000 0x200>,  /* LED regs */
+	      <0x12a200 0x200>,  /* SMI regs */
+	      <0x12a400 0x200>,  /* TAI regs */
+	      <0x12a600 0x200>,  /* XSMI regs */
+	      <0x12b000 0x1000>,  /* MG Internal regs */
+	      <0x130000 0x6000>, /* MSPG regs */
+	      <0x130400 0x200>,  /* MSPG - XPCS regs */
+	      <0x130e00 0x100>,  /* MSPG - GMAC regs */
+	      <0x130f00 0x100>,  /* MSPG - XLG MAC regs */
+	      <0x441000 0x1000>;  /* RFU-1 Regs */
+	reg-names = "pp", "serdes", "xmib", "led", "smi", "tai", "xsmi", "mg", "mspg", "xpcs", "gmac", "xlg", "rfu1";
+	clocks = <&gateclk_1 3>, <&gateclk_1 18>, <&gateclk_1 9>, <&gateclk_1 6>, <&gateclk_1 5>;
+	clock-names = "pp_clk", "gop_core_clk", "gop_clk", "mg_core_clk", "mg_clk";
+	status = "okay";
+	eth0_1: eth0@010000 {
+		interrupts = <GIC_SPI 261 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#0 Interrupt */
+			     <GIC_SPI 265 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#1 Interrupt */
+			     <GIC_SPI 269 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#2 Interrupt */
+			     <GIC_SPI 273 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#3 Interrupt */
+			     <GIC_SPI 277 IRQ_TYPE_LEVEL_HIGH>; /* RX Shared Interrupt */
+		port-id = <0>; /* pp2_port_id */
+		emac-data = <&emac0_1>;
+		status = "disabled";
+	};
+	eth1_1: eth1@020000 {
+		interrupts = <GIC_SPI 262 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#0 Interrupt */
+			     <GIC_SPI 266 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#1 Interrupt */
+			     <GIC_SPI 270 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#2 Interrupt */
+			     <GIC_SPI 274 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#3 Interrupt */
+			     <GIC_SPI 278 IRQ_TYPE_LEVEL_HIGH>; /* RX Shared Interrupt */
+		port-id = <1>; /* pp2_port_id */
+		emac-data = <&emac2_1>;
+		status = "disabled";
+	};
+	eth2_1: eth2@030000 {
+		interrupts = <GIC_SPI 263 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#0 Interrupt */
+			     <GIC_SPI 267 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#1 Interrupt */
+			     <GIC_SPI 271 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#2 Interrupt */
+			     <GIC_SPI 275 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#3 Interrupt */
+			     <GIC_SPI 279 IRQ_TYPE_LEVEL_HIGH>; /* RX Shared Interrupt */
+		port-id = <2>; /* pp2_port_id */
+		emac-data = <&emac3_1>;
+		status = "disabled";
+	};
+};
+eip197@800000 {
+	compatible = "marvell,eip197";
+	reg = <0x800000 0x200000>;
+	dma-coherent;
+	interrupts = <GIC_SPI 278 IRQ_TYPE_LEVEL_HIGH>;
+	clocks = <&gateclk_1 26>;
+	status = "disabled";
+};
diff --git a/arch/arm64/boot/dts/marvell/armada-cp110-slave.dtsi b/arch/arm64/boot/dts/marvell/armada-cp110-slave.dtsi
index 6e20588..fd9a530 100644
--- a/arch/arm64/boot/dts/marvell/armada-cp110-slave.dtsi
+++ b/arch/arm64/boot/dts/marvell/armada-cp110-slave.dtsi
@@ -49,51 +49,20 @@
 		#address-cells = <2>;
 		#size-cells = <2>;
 		compatible = "simple-bus";
-		interrupt-parent = <&gic>;
+		interrupt-parent = <&icu_s>;
 		ranges;
 
 		config-space@f4000000 {
 			#address-cells = <1>;
 			#size-cells = <1>;
 			compatible = "simple-bus";
-			interrupt-parent = <&gic>;
 			ranges = <0x0 0x0 0xf4000000 0x2000000>;
 
 			cps_rtc: rtc@284000 {
 				compatible = "marvell,armada-8k-rtc";
 				reg = <0x284000 0x20>, <0x284080 0x24>;
 				reg-names = "rtc", "rtc-soc";
-				interrupts = <GIC_SPI 71 IRQ_TYPE_LEVEL_HIGH>;
-			};
-
-			cps_ethernet: ethernet@0 {
-				compatible = "marvell,armada-7k-pp22";
-				reg = <0x0 0x100000>, <0x129000 0xb000>;
-				clocks = <&cps_syscon0 1 3>, <&cps_syscon0 1 9>, <&cps_syscon0 1 5>;
-				clock-names = "pp_clk", "gop_clk", "mg_clk";
-				status = "disabled";
-				dma-coherent;
-
-				cps_eth0: eth0 {
-					interrupts = <GIC_SPI 261 IRQ_TYPE_LEVEL_HIGH>;
-					port-id = <0>;
-					gop-port-id = <0>;
-					status = "disabled";
-				};
-
-				cps_eth1: eth1 {
-					interrupts = <GIC_SPI 262 IRQ_TYPE_LEVEL_HIGH>;
-					port-id = <1>;
-					gop-port-id = <2>;
-					status = "disabled";
-				};
-
-				cps_eth2: eth2 {
-					interrupts = <GIC_SPI 263 IRQ_TYPE_LEVEL_HIGH>;
-					port-id = <2>;
-					gop-port-id = <3>;
-					status = "disabled";
-				};
+				interrupts = <ICU_GRP_NSR 77 IRQ_TYPE_LEVEL_HIGH>;
 			};
 
 			cps_mdio: mdio@12a200 {
@@ -103,6 +72,13 @@
 				reg = <0x12a200 0x10>;
 			};
 
+			mdio@15b000 {
+				#address-cells = <1>;
+				#size-cells = <0>;
+				compatible = "marvell,xmdio";
+				reg = <0x12a600 0x10>;
+			};
+
 			cps_syscon0: system-controller@440000 {
 				compatible = "marvell,cp110-system-controller0",
 					     "syscon";
@@ -127,7 +103,7 @@
 				compatible = "marvell,armada-8k-ahci",
 					     "generic-ahci";
 				reg = <0x540000 0x30000>;
-				interrupts = <GIC_SPI 287 IRQ_TYPE_LEVEL_HIGH>;
+				interrupts = <ICU_GRP_NSR 107 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&cps_syscon0 1 15>;
 				status = "disabled";
 			};
@@ -137,7 +113,7 @@
 					     "generic-xhci";
 				reg = <0x500000 0x4000>;
 				dma-coherent;
-				interrupts = <GIC_SPI 286 IRQ_TYPE_LEVEL_HIGH>;
+				interrupts = <ICU_GRP_NSR 106 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&cps_syscon0 1 22>;
 				status = "disabled";
 			};
@@ -147,7 +123,7 @@
 					     "generic-xhci";
 				reg = <0x510000 0x4000>;
 				dma-coherent;
-				interrupts = <GIC_SPI 285 IRQ_TYPE_LEVEL_HIGH>;
+				interrupts = <ICU_GRP_NSR 105 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&cps_syscon0 1 23>;
 				status = "disabled";
 			};
@@ -195,7 +171,7 @@
 				reg = <0x701000 0x20>;
 				#address-cells = <1>;
 				#size-cells = <0>;
-				interrupts = <GIC_SPI 310 IRQ_TYPE_LEVEL_HIGH>;
+				interrupts = <ICU_GRP_NSR 120 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&cps_syscon0 1 21>;
 				status = "disabled";
 			};
@@ -205,7 +181,7 @@
 				reg = <0x701100 0x20>;
 				#address-cells = <1>;
 				#size-cells = <0>;
-				interrupts = <GIC_SPI 311 IRQ_TYPE_LEVEL_HIGH>;
+				interrupts = <ICU_GRP_NSR 120 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&cps_syscon0 1 21>;
 				status = "disabled";
 			};
@@ -213,7 +189,7 @@
 			cps_trng: trng@760000 {
 				compatible = "marvell,armada-8k-rng", "inside-secure,safexcel-eip76";
 				reg = <0x760000 0x7d>;
-				interrupts = <GIC_SPI 312 IRQ_TYPE_LEVEL_HIGH>;
+				interrupts = <ICU_GRP_NSR 95 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&cps_syscon0 1 25>;
 				status = "okay";
 			};
@@ -221,17 +197,220 @@
 			cps_crypto: crypto@800000 {
 				compatible = "inside-secure,safexcel-eip197";
 				reg = <0x800000 0x200000>;
-				interrupts = <GIC_SPI 34 IRQ_TYPE_LEVEL_HIGH>,
-					     <GIC_SPI 278 IRQ_TYPE_LEVEL_HIGH>,
-					     <GIC_SPI 279 IRQ_TYPE_LEVEL_HIGH>,
-					     <GIC_SPI 280 IRQ_TYPE_LEVEL_HIGH>,
-					     <GIC_SPI 281 IRQ_TYPE_LEVEL_HIGH>,
-					     <GIC_SPI 282 IRQ_TYPE_LEVEL_HIGH>;
+				interrupts = <ICU_GRP_NSR 87 IRQ_TYPE_LEVEL_HIGH>,
+					     <ICU_GRP_NSR 88 IRQ_TYPE_LEVEL_HIGH>,
+					     <ICU_GRP_NSR 89 IRQ_TYPE_LEVEL_HIGH>,
+					     <ICU_GRP_NSR 90 IRQ_TYPE_LEVEL_HIGH>,
+					     <ICU_GRP_NSR 91 IRQ_TYPE_LEVEL_HIGH>,
+					     <ICU_GRP_NSR 92 IRQ_TYPE_LEVEL_HIGH>;
 				interrupt-names = "mem", "ring0", "ring1",
 						  "ring2", "ring3", "eip";
 				clocks = <&cps_syscon0 1 26>;
 				status = "disabled";
 			};
+
+			icu_s: interrupt-controller@1E0000 {
+				compatible = "marvell,icu";
+				reg = <0x1E0000 0x10>;
+				#interrupt-cells = <3>;
+				#size-cells = <1>;
+				#address-cells = <0>;
+				icu-spi-base = <0>;
+				interrupt-controller;
+				interrupt-parent = <&gic>;
+				gicp-spi = <0x0 0xF03F0040 0x0 0xF03F0048>;
+			};
+
+			utmi0_1: utmi@580000 {
+				compatible = "marvell,cp110-utmi-phy";
+				reg = <0x580000 0x1000>,
+				      <0x440420 0x4>,
+				      <0x440440 0x4>;
+				reg-names = "utmi", "usb-cfg-reg", "utmi-cfg-reg";
+				utmi-port = <UTMI_PHY_TO_USB3_HOST0>;
+				utmi-index = <0>;
+				#phy-cells = <0>;
+				status = "disabled";
+			};
+
+			nand@720000 {
+				compatible = "marvell,armada370-nand";
+				reg = <0x720000 0x54>;
+				#address-cells = <1>;
+				#size-cells = <1>;
+				num-cs = <1>;
+				marvell,nand-enable-arbiter;
+				nand-on-flash-bbt;
+				interrupts = <ICU_GRP_NSR 115 IRQ_TYPE_LEVEL_HIGH>;
+				clocks = <&cps_syscon0 1 2>;
+				nand-ecc-strength = <4>;
+				nand-ecc-step-size = <512>;
+				status = "disabled";
+			};
+
+			tdm@7a0000 {
+				compatible = "marvell,armada-a8k-tdm";
+				reg = <0x7a0000 0x20000>;
+				reg-names = "tdm_regs";
+				interrupts = <ICU_GRP_NSR 111 IRQ_TYPE_LEVEL_HIGH>,
+					     <ICU_GRP_NSR 112 IRQ_TYPE_LEVEL_HIGH>,
+					     <ICU_GRP_NSR 113 IRQ_TYPE_LEVEL_HIGH>;
+				clocks = <&cps_syscon0 1 1>;
+				clock-names = "gateclk";
+				status = "disabled";
+			};
+
+			thermal@400078 {
+				compatible = "marvell,armada-cp110-thermal";
+				reg = <0x400078 0x4>,
+					<0x400070 0x8>,
+					<0x400100 0x20>;
+				/* interrupt assignment by ICU supports single thermal sensor unit,
+				 * currently cp0 thermal sensor irq is enabled.
+				 * To enable cp1 thermal sensor interrupt,
+				 * first disable cp0 thermal sensor interrupt.
+				 */
+				/* interrupts-extended = <&sei 37>;*/
+				threshold = <100>;
+				hysteresis = <2>;
+				status = "okay";
+			};
+
+			axim-cp-rd@3c5000 {
+				compatible = "marvell,coresight-axim", "arm,primecell";
+				reg = <0x3c5000 0x1000>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+				bus-width = <40>;
+			};
+
+			axim-cp-wr@3c6000 {
+				compatible = "marvell,coresight-axim", "arm,primecell";
+				reg = <0x3c6000 0x1000>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+				bus-width = <40>;
+			};
+
+			axim-ppv2-rd@3c0000 {
+				compatible = "marvell,coresight-axim", "arm,primecell";
+				reg = <0x3c0000 0x1000>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+				bus-width = <40>;
+			};
+
+			axim-ppv2-wr@3c1000 {
+				compatible = "marvell,coresight-axim", "arm,primecell";
+				reg = <0x3c1000 0x1000>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+				bus-width = <40>;
+			};
+
+			axim-hb1-rd@3c8000 {
+				compatible = "marvell,coresight-axim", "arm,primecell";
+				reg = <0x3c8000 0x1000>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+				bus-width = <40>;
+			};
+
+			axim-hb1-wr@3c9000 {
+				compatible = "marvell,coresight-axim", "arm,primecell";
+				reg = <0x3c9000 0x1000>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+				bus-width = <40>;
+			};
+
+			gop {
+				emac0_1: mac0 {
+					interrupts = <ICU_GRP_NSR 129 IRQ_TYPE_LEVEL_HIGH>;  /* Link IRQ */
+					mac-address = [00 00 00 00 00 04];
+					port-id = <0>; /* gop_port_id */
+				};
+				emac2_1: mac2 {
+					interrupts = <ICU_GRP_NSR 128 IRQ_TYPE_LEVEL_HIGH>;  /* Link IRQ */
+					mac-address = [00 00 00 00 00 05];
+					port-id = <2>; /* gop_port_id */
+				};
+				emac3_1: mac3 {
+					interrupts = <ICU_GRP_NSR 127 IRQ_TYPE_LEVEL_HIGH>;  /* Link IRQ */
+					mac-address = [00 00 00 00 00 06];
+					port-id = <3>; /* gop_port_id */
+				};
+			};
+
+			ppv22@000000 {
+				compatible = "marvell,mv-pp22";
+				reg = <0x000000 0x90000>, /* Packet Processor regs */
+				      <0x129000 0x0600>, /* XMIB regs */
+				      <0x12a000 0x200>,  /* LED regs */
+				      <0x12a200 0x200>,  /* SMI regs */
+				      <0x12a400 0x200>,  /* TAI regs */
+				      <0x12a600 0x200>,  /* XSMI regs */
+				      <0x12b000 0x1000>,  /* MG Internal regs */
+				      <0x130000 0x6000>, /* MSPG regs */
+				      <0x130400 0x200>,  /* MSPG - XPCS regs */
+				      <0x130600 0x200>,  /* FCA - flow control regs*/
+				      <0x130e00 0x100>,  /* MSPG - GMAC regs */
+				      <0x130f00 0x100>,  /* MSPG - XLG MAC regs */
+				      <0x441100 0x100>;  /* RFU-1 Regs */
+				reg-names = "pp", "xmib", "led", "smi", "tai", "xsmi", "mg", "mspg", "xpcs",
+					    "fca", "gmac", "xlg", "rfu1";
+				clocks = <&cps_syscon0 1 3>, <&cps_syscon0 1 18>, <&cps_syscon0 1 9>,
+					 <&cps_syscon0 1 6>, <&cps_syscon0 1 5>;
+				clock-names = "pp_clk", "gop_core_clk", "gop_clk", "mg_core_clk", "mg_clk";
+				status = "okay";
+				eth0_1: eth0@010000 {
+					interrupts = <ICU_GRP_NSR 39 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#0 Interrupt */
+						     <ICU_GRP_NSR 43 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#1 Interrupt */
+						     <ICU_GRP_NSR 47 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#2 Interrupt */
+						     <ICU_GRP_NSR 51 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#3 Interrupt */
+						     <ICU_GRP_NSR 55 IRQ_TYPE_LEVEL_HIGH>; /* RX Shared Interrupt */
+					port-id = <0>; /* pp2_port_id */
+					emac-data = <&emac0_1>;
+					status = "disabled";
+				};
+				eth1_1: eth1@020000 {
+					interrupts = <ICU_GRP_NSR 40 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#0 Interrupt */
+						     <ICU_GRP_NSR 44 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#1 Interrupt */
+						     <ICU_GRP_NSR 48 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#2 Interrupt */
+						     <ICU_GRP_NSR 52 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#3 Interrupt */
+						     <ICU_GRP_NSR 56 IRQ_TYPE_LEVEL_HIGH>; /* RX Shared Interrupt */
+					port-id = <1>; /* pp2_port_id */
+					emac-data = <&emac2_1>;
+					status = "disabled";
+				};
+				eth2_1: eth2@030000 {
+					interrupts = <ICU_GRP_NSR 41 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#0 Interrupt */
+						     <ICU_GRP_NSR 45 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#1 Interrupt */
+						     <ICU_GRP_NSR 49 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#2 Interrupt */
+						     <ICU_GRP_NSR 53 IRQ_TYPE_LEVEL_HIGH>, /* TX CPU#3 Interrupt */
+						     <ICU_GRP_NSR 57 IRQ_TYPE_LEVEL_HIGH>; /* RX Shared Interrupt */
+					port-id = <2>; /* pp2_port_id */
+					emac-data = <&emac3_1>;
+					status = "disabled";
+				};
+			};
+
+			comphy1: comphy {
+				compatible = "marvell,cp110-comphy";
+				reg = <0x441000 0x100>, <0x120000 0x6000>;
+				reg-names = "comphy", "serdes";
+				dfx-reg = <0x400280 4>;
+				#phy-cells = <2>;
+				sar-data = <&cps_sar>;
+				skip_pcie_power_off;
+				status = "disabled";
+			};
+
+			cps_sar: sar@400200 {
+				compatible = "marvell,sample-at-reset-cp110";
+				reg = <0x400200 0x8>;
+				status = "okay";
+			};
 		};
 
 		cps_pcie0: pcie@f4600000 {
@@ -253,8 +432,8 @@
 				/* non-prefetchable memory */
 				0x82000000 0 0xfa000000 0  0xfa000000 0 0xf00000>;
 			interrupt-map-mask = <0 0 0 0>;
-			interrupt-map = <0 0 0 0 &gic 0 GIC_SPI 256 IRQ_TYPE_LEVEL_HIGH>;
-			interrupts = <GIC_SPI 256 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-map = <0 0 0 0 &icu_s 0 ICU_GRP_NSR 22 IRQ_TYPE_LEVEL_HIGH>;
+			interrupts = <ICU_GRP_NSR 22 IRQ_TYPE_LEVEL_HIGH>;
 			num-lanes = <1>;
 			clocks = <&cps_syscon0 1 13>;
 			status = "disabled";
@@ -279,8 +458,8 @@
 				/* non-prefetchable memory */
 				0x82000000 0 0xfb000000 0  0xfb000000 0 0xf00000>;
 			interrupt-map-mask = <0 0 0 0>;
-			interrupt-map = <0 0 0 0 &gic 0 GIC_SPI 258 IRQ_TYPE_LEVEL_HIGH>;
-			interrupts = <GIC_SPI 258 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-map = <0 0 0 0 &icu_s 0 ICU_GRP_NSR 24 IRQ_TYPE_LEVEL_HIGH>;
+			interrupts = <ICU_GRP_NSR 24 IRQ_TYPE_LEVEL_HIGH>;
 
 			num-lanes = <1>;
 			clocks = <&cps_syscon0 1 11>;
@@ -306,8 +485,8 @@
 				/* non-prefetchable memory */
 				0x82000000 0 0xfc000000 0  0xfc000000 0 0xf00000>;
 			interrupt-map-mask = <0 0 0 0>;
-			interrupt-map = <0 0 0 0 &gic 0 GIC_SPI 257 IRQ_TYPE_LEVEL_HIGH>;
-			interrupts = <GIC_SPI 257 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-map = <0 0 0 0 &icu_s 0 ICU_GRP_NSR 23 IRQ_TYPE_LEVEL_HIGH>;
+			interrupts = <ICU_GRP_NSR 23 IRQ_TYPE_LEVEL_HIGH>;
 
 			num-lanes = <1>;
 			clocks = <&cps_syscon0 1 12>;
-- 
1.7.9.5

